### PR TITLE
SONARJAVA-4832 Implement S6881, Fix SONARJAVA-4913

### PR DESCRIPTION
**Pull request by johann-beleites-sonarsource**
VirtualThreads should be used for tasks that include heavy blocking operations.

### CODE REVIEW

**Comment by johann-beleites-sonarsource** [comment]
Please rebase, don't squash.

**Review by irina-batinic-sonarsource — COMMENTED** [review]
The implementation of the rule looks good, as does the fix for S1181. Good job!
I left a comment about the rule description on [this PR](https://github.com/SonarSource/rspec/pull/3780).

File: sonar-java-plugin/src/main/resources/org/sonar/l10n/java/rules/java/S6881.html [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+static void processRequest(String data) {
+    try {
+        // do something with data
+        Thread.sleep(1000);
+    } catch (InterruptedException e) {
+        e.printStackTrace();
+    }
+}
+</pre>
```
Comment (irina-batinic-sonarsource) [inline sonar-java-plugin/src/main/resources/org/sonar/l10n/java/rules/java/S6881.html:36]: I wondered why we have this Noncompliant code example without a corresponding Compliant Solution.

File: java-checks/src/main/java/org/sonar/java/checks/BlockingOperationsInVirtualThreadsCheck.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    .ofSubTypes(JAVA_LANG_THREAD)
+    .names("run")
+    .addWithoutParametersMatcher()
+    .build();
+
+  @Override
+  public void visitNode(Tree tree) {
+    switch (tree.kind()) {
+      case NEW_CLASS -> onConstructorFound((NewClassTree) tree);
+      case METHOD -> onMethodFound((MethodTree) tree);
```
Comment (leonardo-pilastri-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/BlockingOperationsInVirtualThreadsCheck.java:95]:
```suggestion
      case METHOD -> onMethodFound((MethodTree) tree);
      case METHOD_INVOCATION -> onMethodInvocationFound((MethodInvocationTree) tree);
```

File: java-checks/src/main/java/org/sonar/java/checks/BlockingOperationsInVirtualThreadsCheck.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      analyzeForIssues(newClassTree.arguments().get(0), newClassTree.identifier());
+    }
+  }
+
+  private void onMethodFound(MethodTree tree) {
+    if (overwritableThreadMethods.matches(tree)) {
+      analyzeForIssues(tree, tree.simpleName());
+    }
+  }
+
```
Comment (leonardo-pilastri-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/BlockingOperationsInVirtualThreadsCheck.java:110]: And with this fairly simple implementation we should be able to cover the extra constructor!
```

  private void onMethodInvocationFound(MethodInvocationTree tree) {
    var builderType = ((MemberSelectExpressionTree) tree.methodSelect()).expression().symbolType();
    if (threadCreationConstructors.matches(tree) && builderType.is("java.lang.Thread$Builder$OfPlatform")) {
      analyzeForIssues(tree.arguments().get(0), tree.methodSelect());
    }
  }
```

**Review by leonardo-pilastri-sonarsource — CHANGES_REQUESTED** [review]
Looks good! I think we can easily cover an extra way to construct a platform thread, then I think the PR is good to go!

**Review by leonardo-pilastri-sonarsource — APPROVED** [review]
Nice work!

**Pull request merged**